demos/device-shadow-for-aws-iot-embedded-sdk/shadow_demo_main.c

```diff
@@ -0,0 +1,738 @@
+
+/* shadow demo helpers header. */
+#include "shadow_demo_helpers.h"
+
+
+/**
+ * @brief Format string representing a Shadow document with a "desired" state.
+ *
+ * The real json document will look like this:
+ * {
+ *   "state": {
+ *     "desired": {
+ *       "powerOn": 1
+ *     }
+ *   },
+ *   "clientToken": "021909"
+ * }
+ *
+ * Note the client token, which is optional for all Shadow updates. The client
+ * token must be unique at any given time, but may be reused once the update is
+ * completed. For this demo, a timestamp is used for a client token.
+ */
+#define SHADOW_DESIRED_JSON     \
+    "{"                         \
+    "\"state\":{"               \
+    "\"desired\":{"             \
+    "\"powerOn\":%01d"          \
+    "}"                         \
+    "},"                        \
+    "\"clientToken\":\"%06lu\"" \
+    "}"
+
+/**
+ * @brief The expected size of #SHADOW_DESIRED_JSON.
+ *
+ * Because all the format specifiers in #SHADOW_DESIRED_JSON include a length,
+ * its full actual size is known by pre-calculation, here's the formula why
+ * the length need to minus 3:
+ * 1. The length of "%01d" is 4.
+ * 2. The length of %06lu is 5.
+ * 3. The actual length we will use in case 1. is 1 ( for the state of powerOn ).
+ * 4. The actual length we will use in case 2. is 6 ( for the clientToken length ).
+ * 5. Thus the additional size 3 = 4 + 5 - 1 - 6 + 1 (termination character).
+ *
+ * In your own application, you could calculate the size of the json doc in this way.
+ */
+#define SHADOW_DESIRED_JSON_LENGTH    ( sizeof( SHADOW_DESIRED_JSON ) - 3 )
+
+/**
+ * @brief Format string representing a Shadow document with a "reported" state.
+ *
+ * The real json document will look like this:
+ * {
+ *   "state": {
+ *     "reported": {
+ *       "powerOn": 1
+ *     }
+ *   },
+ *   "clientToken": "021909"
+ * }
+ *
+ * Note the client token, which is required for all Shadow updates. The client
+ * token must be unique at any given time, but may be reused once the update is
+ * completed. For this demo, a timestamp is used for a client token.
+ */
+#define SHADOW_REPORTED_JSON    \
+    "{"                         \
+    "\"state\":{"               \
+    "\"reported\":{"            \
+    "\"powerOn\":%01d"          \
+    "}"                         \
+    "},"                        \
+    "\"clientToken\":\"%06lu\"" \
+    "}"
+
+/**
+ * @brief The expected size of #SHADOW_REPORTED_JSON.
+ *
+ * Because all the format specifiers in #SHADOW_REPORTED_JSON include a length,
+ * its full size is known at compile-time by pre-calculation. Users could refer to
+ * the way how to calculate the actual length in #SHADOW_DESIRED_JSON_LENGTH.
+ */
+#define SHADOW_REPORTED_JSON_LENGTH    ( sizeof( SHADOW_REPORTED_JSON ) - 3 )
+
+#ifndef THING_NAME
+
+/**
+ * @brief Predefined thing name.
+ *
+ * This is the example predefine thing name and could be compiled in ROM code.
+ */
+    #define THING_NAME    clientcredentialIOT_THING_NAME
```

Comment:
can we change this to democonfigCLIENT_IDENTIFIER for some consistency
```suggestion
    #define THING_NAME   democonfigCLIENT_IDENTIFIER
```

Reply:
sure.

Reply:
please check this [commit](https://github.com/aws/amazon-freertos/pull/2537/commits/09f117053a734f47e94418c012902c28fdc3cfe2).

Reply:
Yea but, `democonfigCLIENT_IDENTIFIER` is intended to be configured in the shadow_demo_config.h by the customer, so I think we can move lines 60 to 93 in shadow_demo_helpers.c to shadow_demo_helpers.h to share the configs.